Alternative proposal for Actor PubSub implementation

WhitWaldo opened the pull request:
None of the comments have been addressed in the various reviews on the [existing Actors PubSub proposal](https://github.com/dapr/proposals/pull/74) nor has any discussion occurred on the thread I created in the maintainers channel on Discord.

But as there is voting occurring on the existing proposal, I'm making a last-ditch effort to suggest ditching that proposal in favor of an alternative that I think is more broadly applicable to what people want to get out of Actors and PubSub, repurposes what's already available and used in Dapr and avoids having additional "one component for all occasions" dependencies.

Fundamentally, I take what we already use for Dapr PubSub today (declarative, programmatic and streaming subscriptions) and extend it to Actors. Contrary to the existing proposal which changes how anyone would have to send events in to use them in actors, my proposal allows actors to subscribe to any existing PubSub component without changing how the events come in and only changing how actors themselves subscribe. In other words, mine is a drop-in change that anyone can use on day-one with existing PubSub without needing anything more than annotating actor methods with corresponding routing (for declarative or programmatic subscriptions).

This is a replacement for the PR at #75 since I accidentally created it from the wrong branch

I welcome any and all feedback!

theperm commented:
I like this approach. Thumbs up for CE subscriptions spec - thats only just come to my attention but make sense for this.

cicoyle commented on 20250322-actors-pubsub.md:
```diff
@@ -0,0 +1,367 @@
+subscription so that there is no substantive difference in how one subscription is registered versus another, lessening
+the burden on the runtime to juggle different approaches.
+
+#### Streaming Subscriptions API
```
Are we going to implement this in a phased approach? Like start with streaming subscriptions then add to declarative and programmatic? What is the thought process here for scope?

WhitWaldo replied:
I think the result of our chat Monday was that we should look at supporting streaming subscriptions first and only revisit with the remaining options later if there's demand for it. Declarative supports hot reload, but that requires extra-code changes to make and programmatic is fixed at app load (and is finnicky with actors because of the activation requirement), so having an intentional opt-in at the actor instance level is the most straightforward route.

I'll modify the proposal itself to reflect this in the next day or two.

cicoyle reviewed (commented):
I generally support the proposal. Just 2 clarifications

olitomlinson commented:
> I cannot think of a PubSub provider that supports more than 32 consumer groups

PubSub components that don't have a concept of partitions, could theoretically have an unbounded amount of consumer groups.

But, I agree, consumer groups are not the right API to target specific Actors with.

olitomlinson commented:
+1 for this alternate proposal for the BYO PubSub component aspect.

Although I don't fully understand the **Subscription Manager** concept. What I actually mean is I don't see how you intend to implement this. How do you go from a single PubSub message, to reliably delivering that to many (potentially unbounded amount of...) Actors, across N number of pods.

WhitWaldo commented:
> +1 for this alternate proposal for the BYO PubSub component aspect.
> 
> Although I don't fully understand the **Subscription Manager** concept. What I actually mean is I don't see how you intend to implement this. How do you go from a single PubSub message, to reliably delivering that to many (potentially unbounded amount of...) Actors, across N number of pods.

The idea for the subscription manager is more just the notion that there's a dedicated capability in the runtime that's managing subscriptions (potentially for deduplication purposes (though that's fraught with concern for later use cases) or other optimization opportunities) and to provide a sample of what the schema for those objects looks like consistent with the CloudEvent subscription spec (as we're using CloudEvents anyway) - I'll update the proposal to reflect this preference in the next day or two.

Where today's PubSub leaves it to the subscriber to handle ack/drop/retry policy, I don't know how scalable that is here where you might have 10 or 10 million actor instances with separate subscriptions. My proposal instead suggests turning that on its head with regards to actors to say that Dapr itself receives the message and attempts to resolve somewhere to send it to (per any registered subscriptions to that component/topic and then of those, which match filters) and if it finds a destination, it acknowledges the message to the broker. If not, it signals a retry or drop based on policy. From there, handling resiliency between the runtime and the clients is up to resiliency policies set up.

WhitWaldo commented on 20250322-actors-pubsub.md:
```diff
@@ -0,0 +1,367 @@
+
+// The message containing the details for subscribing an actor to a topic via streaming or otherwise
+message SubscribeActorEventRequestAlpha1 {
+    oneof subscribe_topic_events_request_type {
```
I've gone through and reworked the entire HTTP and gRPC specification for each of the management operations (create, get, update, delete, query), so this has since been removed in favor of something that made more sense and was specific to my proposal.

WhitWaldo commented:
I've made some significant changes to the text of the proposal but centering around three themes that came up in the last two weeks of discussions about the idea:
- Dropped initial support for declarative and programmatic subscriptions in favor of streaming subscriptions
- Replaced the blend of CE and CEL filtering with a pure CE subscription spec implementation favoring the latest WIP and published CloudEvents specifications instead of having us reinvent the wheel with a funky hybrid.
- The protos as originally provided didn't quite align with the concept as I imagined it, so I've added an HTTP specification and completely revamped the corresponding gRPC specification to go along with it (matching the new CE specification)

msfussell commented:
I always find it useful if you can provide at least one SDK language (preferrably two) that show the end user code example. For example .NET and Python examples of how an Actor subscribes to a message. I realize that this is SDK dependent and may not be exactly write, but it gives the idea on how to use the Dapr runtime API.

WhitWaldo commented:
> I always find it useful if you can provide at least one SDK language (preferrably two) that show the end user code example. For example .NET and Python examples of how an Actor subscribes to a message. I realize that this is SDK dependent and may not be exactly write, but it gives the idea on how to use the Dapr runtime API.

That's sort of the beauty of this proposal, but one that Josh and I have been tackling offline is that there's potentially no changes needed to any of the actor implementations to support this because it's just like firing a timer - you're effectively specifying a method name that'll be invoked with the payload when a message matches up with that actor ID.

The only thing you'd need is the mechanism for scheduling the workflow and in C#, that might look very similar to any other Dapr method:

```cs
public sealed record DaprPubSubSink(string ActorType, string ActorMethod, string? ActorId = null)
    {
        public override string ToString()
        {
            var value = $"{ActorType}/{ActorMethod}";
            if (ActorId is not null)
                value += $"/{ActorId}";
            return value;
        }
    }

    public abstract Task<string> RegisterActorPubSubSubscriptionAsync(string pubsubName, string topicName, DaprPubSubSink sink, string? subscriptionId = null, IReadOnlyList<string>? types = null, IReadOnlyList<DaprSubscriptionFilter>? filters = null, CancellationToken cancellationToken = default);
```

WhitWaldo commented on 20250322-actors-pubsub.md:
```diff
@@ -0,0 +1,547 @@
+any subscribers. The publisher does not know who is subscribing to the message(s) and it's very possible that there's
+more than one subscriber across actor types and instances allowing more than one actor to react to the same message.
+
+This does not seek to implement actor message passing or fire-and-forget asynchronous communication. While valuable
```
What I was trying to get at in a larger sense is that this isn't seeking to enable a broad, developer-accessible fire and forget capability even if that's what is practically done in this circumstance.

WhitWaldo commented on 20250322-actors-pubsub.md:
```diff
@@ -0,0 +1,547 @@
+    // Required: The name of the PubSub topic/queue the event is sourced from
+    "topic": "<pubsub_topic_name>"
+  },
+  // Optional: If specified, an array of at least one ilter expression that evaluates to true or false. Delivery 
```
Fixed

WhitWaldo commented:
On the release call this week, we elected to push this back to the 1.17 release so it can get the development time it deserves and I agree with this decision.
